server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -523,26 +524,16 @@ protected void doRun() {
                         throw new IllegalArgumentException("only write ops with an op_type of create are allowed in data streams");
                     }
 
-                    IndexRouting indexRouting = concreteIndices.routing(concreteIndex);
-
-                    switch (docWriteRequest.opType()) {
-                        case CREATE:
-                        case INDEX:
-                            prohibitAppendWritesInBackingIndices(docWriteRequest, metadata);
-                            prohibitCustomRoutingOnDataStream(docWriteRequest, metadata);
-                            IndexRequest indexRequest = (IndexRequest) docWriteRequest;
-                            indexRequest.resolveRouting(metadata);
-                            indexRequest.process();
-                            break;
-                        case UPDATE:
-                            docWriteRequest.routing(metadata.resolveWriteIndexRouting(docWriteRequest.routing(), docWriteRequest.index()));
-                            break;
-                        case DELETE:
-                            docWriteRequest.routing(metadata.resolveWriteIndexRouting(docWriteRequest.routing(), docWriteRequest.index()));
-                            break;
-                        default:
-                            throw new AssertionError("request type not supported: [" + docWriteRequest.opType() + "]");
+                    if (docWriteRequest.opType() == OpType.CREATE || docWriteRequest.opType() == OpType.INDEX) {
+                        prohibitAppendWritesInBackingIndices(docWriteRequest, metadata);
+                        prohibitCustomRoutingOnDataStream(docWriteRequest, metadata);
```

Comment:
@martijnvg I wonder if it is not a bug that we do not check this for all request types, including update/delete? Maybe better handled in a follow-up, but would be nice to move out of this "if". And perhaps then let `prohibitAppendWritesInBackingIndices` handle the ops itself to simplify this code (the filter for CREATE/INDEX here is really only partial anyway, so might as well do that inside the method).

Reply:
Yes, I think `prohibitCustomRoutingOnDataStream()` should be checked for all types of request, irregardless whether these target the data stream of one of the backing indices. I think only `prohibitAppendWritesInBackingIndices()` should be checked in this if statement. I can work on that in a followup.

Reply:
Neat! I was hoping writing it like this would make stuff like this more obvious. I'll leave this to @martijnvg's follow up though.